onnxruntime/core/optimizer/matmul_scale_fusion.cc

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/matmul_scale_fusion.h"
+
+#include "onnx/defs/attr_proto_util.h"
+
+#include "core/common/optional.h"
+#include "core/framework/data_types_internal.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/graph/graph_utils.h"
+#include "core/graph/graph_viewer.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+
+namespace {
+template <typename T>
+struct ExtractScalarAsFloatDispatchTarget {
+  Status operator()(const ONNX_NAMESPACE::TensorProto& tensor_proto, float& scalar_float) {
+    T scalar;
+    ORT_RETURN_IF_ERROR(utils::UnpackTensor(tensor_proto, &scalar, 1));
+    scalar_float = static_cast<float>(scalar);
+    return Status::OK();
+  }
+};
+
+optional<float> GetScalarConstantInitializer(const Graph& graph, const NodeArg& node_arg) {
+  const auto* initializer = graph_utils::GetConstantInitializer(graph, node_arg.Name());
+
+  if (!initializer) {
+    // not a constant
+    return {};
+  }
+
+  const auto* shape = node_arg.Shape();
+  ORT_ENFORCE(
+      shape,
+      "Constant initializer NodeArg shape should not be null. NodeArg: ", node_arg.Name());
+
+  if (utils::GetTensorShapeFromTensorShapeProto(*shape).Size() != 1) {
+    // not a scalar
+    return {};
+  }
```

Comment:
we have two shapes to represent a scaler: {1} and {}

Reply:
i think TensorShape::Size() == 1 should account for both of them, as well as other stuff like {1, 1, 1}. is that too lenient?

---
In reply to: [464692344](https://github.com/microsoft/onnxruntime/pull/4669#discussion_r464692344) [](ancestors = 464692344)

Reply:
great! 

---
In reply to: [464693290](https://github.com/microsoft/onnxruntime/pull/4669#discussion_r464693290) [](ancestors = 464693290,464692344)